Add alternative summary UI with arrow icon to the Disclosure component

digiwand opened the pull request:
## Explanation

Fixes: https://github.com/MetaMask/metamask-extension/issues/19768
Blocks: https://github.com/MetaMask/MetaMask-planning/issues/558

Adds alternative `<summary>`  UI to the Disclosure component. See details in issue ticket above.

In this PR: 
- add `type` prop to change from the default `<summary>`  UI and the arrow `<summary>`  UI
- alphabetize props
- add `cursor: pointer` to actionable `<summary>` element
- Arrow Type specific details that differ from default:
  - no margin/padding left on content
  - support responsive `<summary>`  text
  - blue text
  - icon on right

## Screenshots/Screencaps

<!-- If you're making a change to the UI, make sure to capture a screenshot or a short video showing off your work! -->

### Closed
<img width="911" alt="Screen Shot 2023-06-28 at 6 23 36 PM" src="https://github.com/MetaMask/metamask-extension/assets/20778143/d7ac7f72-67b8-4790-88d5-2606667b742e">

### Opened
<img width="918" alt="Screen Shot 2023-06-28 at 6 23 45 PM" src="https://github.com/MetaMask/metamask-extension/assets/20778143/031d9459-40ad-4996-be7a-8d5e19892314">

### Disclosure with responsive text size (large)
<img width="794" alt="Screen Shot 2023-07-10 at 9 15 35 PM" src="https://github.com/MetaMask/metamask-extension/assets/20778143/d0c50c94-cbad-4c11-82b4-0ad68eb75256">


### Disclosure with responsive text size (small)
<img width="509" alt="Screen Shot 2023-07-10 at 9 15 06 PM" src="https://github.com/MetaMask/metamask-extension/assets/20778143/957eeb96-3f40-4749-9e20-9cb1c901bb99">


<!-- How does it look now? Drag your file(s) below this line: -->

## Manual Testing Steps

<!--
How should reviewers and QA manually test your changes? For instance:

- Go to this screen
- Do this
- Then do this
-->

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [x] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [x] PR is linked to the appropriate GitHub issue
- [x] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

digiwand committed:
disclosure: add isArrowSummary option w/ UI

digiwand committed:
disclosure.scss: add missing trailing semicolon

georgewrmarshall reviewed (dismissed):
LGTM! You could make this component a little more scalable by providing an enum instead of a boolean for the different types? But I don't think it's blocking if the Accordion is on the design system team roadmap and this will be eventually replaced
- checked controls and functionality in storybook

digiwand committed:
Disclosure: replace isArrowSummary -> type prop

digiwand committed:
Disclosure: alphabetize props

digiwand committed:
Disclosure: add cursor: pointer

digiwand commented:
Thanks for the reviews @georgewrmarshall & @NidhiKJha! 

I love the enum proposal and think it may be easier to scale, or convert the future Accordion component, by adding it now rather than later. Made 3 changes since the last reviews:
1. replace `isArrowSummary` with `type` prop
2. alphabetize props
3. add `cursor: pointer` to actionable summary element

may I get a re-review please?

digiwand committed:
Merge branch 'develop' into ds-disclosure-add-isArrowSummary-option

NidhiKJha reviewed (dismissed):
LGTM 🚀

digiwand committed:
storybook: add Disclosure - Type Arrow

digiwand commented:
sorry @NidhiKJha. one more please? 🥹👉🏼👈🏼 I added a storybook page for the type

digiwand committed:
Disclosure: margin-left: 0 arrow type content

digiwand committed:
Disclosure: fix margin-left: 0 arrow type

digiwand commented:
@georgewrmarshall @NidhiKJha 

I added a storybook page for the new type. 
I also realized that for the design we'd like to use, we will not have margin/padding on the content. I updated the style for the arrow type to reflect this. My thought was to not introduce more variants and complexities, e.g. custom padding, until they are needed. Does this sound good to you two? 

<img width="918" alt="Screen Shot 2023-06-28 at 6 23 45 PM" src="https://github.com/MetaMask/metamask-extension/assets/20778143/582e0a79-f66f-4e3b-8d85-0bba31f0bed8">

digiwand commented:
One other thing to address is how to match the text sizing with the BannerAlert modal since it changes depending on the width of the screen. Planning to further look into this.

![Screen Shot 2023-06-29 at 3 52 09 PM](https://github.com/MetaMask/metamask-extension/assets/20778143/550edb25-7ded-4fb2-821e-e30dd539ab0e)

digiwand committed:
Disclosure: support Text media query for Arrow

brad-decker reviewed (dismissed):
LGTM

digiwand committed:
Merge branch 'develop' into ds-disclosure-add-isArrowSummary-option

georgewrmarshall commented on ui/components/ui/disclosure/disclosure.constants.ts:
```diff
@@ -0,0 +1,4 @@
+export enum DISCLOSURE_TYPES {
+  DEFAULT = 'default',
+  ARROW = 'arrow',
+}
```
I believe our enum conventions are to use PascalCase and singular and possibly best not to use the word types as an enum can also be used as a type

https://www.notion.so/Use-TypeScript-9ec7bd8fc8cf4d82b81cd8943c4707cd?pvs=4#336ec1a7c32f44feb02cc8f14727620d

```suggestion
export enum DisclosureVariant {
  Default = 'default',
  Arrow = 'arrow',
}
```

georgewrmarshall commented on ui/components/ui/disclosure/disclosure.js:
```diff
@@ -1,9 +1,48 @@
+ * @param {string} title
+ * @returns {JSX.Element}
+ */
+const renderSummaryByType = (type, title) => {
```
Can we change type to variant?
```suggestion
const renderSummaryByType = (variant, title) => {
```

georgewrmarshall commented on ui/components/ui/disclosure/disclosure.scss:
```diff
@@ -23,4 +24,31 @@
       font-size: 12px;
     }
```
Could we replace these static font sizes with a `Text` component?

digiwand replied:
@georgewrmarshall 

hmm this one might be a little tricky since the content may be elements that are not compatible with the text variants. e.g. It may be a `<div>` which is incompatible as a child of `<p>`

the way I was going about it for the new SecurityProviderBannerAlert was  passing the Text elements as children

ref: https://github.com/MetaMask/metamask-extension/pull/19825/files
see `ui/components/app/security-provider-banner-alert/security-provider-banner-alert.js`

georgewrmarshall replied:
I see. You could also change the root HTML element that is being rendered using the `<Text as="div"` prop. That way the typographic styles are attached using the className on the Text component but the underlying element is a div.

digiwand replied:
I like this suggestion. Ty! Will add to another PR shortly 👍

georgewrmarshall reviewed (commented):
Left one suggestion regarding enum conventions that I missed in my initial review. But otherwise LGTM

digiwand committed:
Disclosure: clean rn type -> variant

digiwand committed:
Merge branch 'develop' into ds-disclosure-add-isArrowSummary-option

digiwand committed:
Disclosure: use size prop for arrow summary text

georgewrmarshall commented on ui/components/ui/disclosure/disclosure.js:
```diff
@@ -1,9 +1,53 @@
+  }
+};
+
+const Disclosure = ({ children, title, size, type }) => {
```
I was also wondering if it made sense to change the `type` prop to `variant` then we have this pattern of `ComponentName` + `PropName` for the enum? Sorry if I was unclear on my last review

digiwand replied:
whoops missed those. updated. thank you! [4e929c7](https://github.com/MetaMask/metamask-extension/pull/19770/commits/4e929c77ad026928625d1dd40940e4ce447b2dfe) cc: @georgewrmarshall

georgewrmarshall reviewed (commented):
Sorry, I may have been unclear in my last review I think it would be great if the `type` prop was also updated to `variant`. Otherwise LGTM

digiwand committed:
Disclosure: update rest of type -> variant

digiwand committed:
Disclosure: 1 more type -> variant

digiwand committed:
Disclosure: TypeArrow -> VariantArrow

cryptotavares reviewed (approved):
LGTM

georgewrmarshall reviewed (approved):
LGTM 🚀
